runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -85,45 +123,83 @@ public void startBundle() {
     doFnRunner.startBundle();
   }
 
+  @Override
+  public void finishBundle() {
+    doFnRunner.finishBundle();
+  }
+
   @Override
   public void processElement(WindowedValue<InputT> input) {
 
     // StatefulDoFnRunner always observes windows, so we need to explode
     for (WindowedValue<InputT> value : input.explodeWindows()) {
-
       BoundedWindow window = value.getWindows().iterator().next();
-
       if (isLate(window)) {
         // The element is too late for this window.
-        droppedDueToLateness.inc();
-        WindowTracing.debug(
-            "StatefulDoFnRunner.processElement: Dropping element at {}; window:{} "
-                + "since too far behind inputWatermark:{}",
-            input.getTimestamp(),
-            window,
-            cleanupTimer.currentInputWatermarkTime());
+        reportDroppedElement(value, window);
+      } else if (requiresTimeSortedInput) {
+        processElementOrdered(window, value);
```

Comment:
This implies that we have to buffer elements up to watermark + allowed lateness.

Reply:
Fixed.